File: example-crs-webservice/crs-userspace/libs/libMSA/java/libmsa/src/test/java/com/teamatlanta/libmsa/thread/ThreadPoolTest.java
package com.teamatlanta.libmsa.thread;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ThreadPoolTest {
    private ThreadPool<Integer> pool;

    @BeforeEach
    public void setUp() {
        pool = null;
    }

    // Mock function that simulates a processing time of 0.0001 seconds
    private void mockFunction(Integer data) {
        try {
            TimeUnit.MICROSECONDS.sleep(100);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    @Test
    public void testBroadcastEnqueue() {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.BROADCAST, this::mockFunction);

        assertEquals(numThreads, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            assertEquals(numData, queue.size());
        }
    }
    
    @Test
    public void testGlobalEnqueue() {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.GLOBAL, this::mockFunction);

        assertEquals(1, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        assertEquals(numData, pool.getWorkQueues()[0].size());
    }

    @Test
    public void testRoundRobinEnqueue() {
        int numThreads = 3;
        int numData = 3000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.ROUND_ROBIN, this::mockFunction);

        assertEquals(numThreads, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            assertEquals(numData / numThreads, queue.size());
        }
    }

    @Test
    public void testBroadcastDequeue() throws InterruptedException {
        int numRepetitions = 2;
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.BROADCAST, this::mockFunction);

        pool.execute();

        for (int t = 0; t < numRepetitions; t++){
            for (int i = 0; i < numData; i++) {
                pool.enqueue(i);
            }
    
            TimeUnit.SECONDS.sleep(3);
    
            for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
                assertEquals(0, queue.size());
            }
        }
    }

    @Test
    public void testGlobalDequeue() throws InterruptedException {
        int numRepetitions = 2;
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.GLOBAL, this::mockFunction);

        pool.execute();

        for (int t = 0; t < numRepetitions; t++){ 
            for (int i = 0; i < numData; i++) {
                pool.enqueue(i);
            }
    
            TimeUnit.SECONDS.sleep(3);
    
            assertEquals(0, pool.getWorkQueues()[0].size());
        }
    }

    @Test
    public void testRoundRobinDequeue() throws InterruptedException {
        int numRepetitions = 2;
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.ROUND_ROBIN, this::mockFunction);

        pool.execute();

        for (int t = 0; t < numRepetitions; t++){ 
            for (int i = 0; i < numData; i++) {
                pool.enqueue(i);
            }
    
            TimeUnit.SECONDS.sleep(3);
    
            for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
                assertEquals(0, queue.size());
            }
        }
    }
    
    @Test
    public void testBroadcastEnqueueAdditionalThreads() {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.BROADCAST, this::mockFunction);

        assertEquals(numThreads, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(numThreads + additionalThreads, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        int cnt1 = 0;
        int cnt2 = 0;

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            if (queue.size() == numData + additionalData) {
                cnt1++;
            } else if (queue.size() == additionalData) {
                cnt2++;
            }
        }

        assertEquals(cnt1, numThreads);
        assertEquals(cnt2, additionalThreads);
    }

    @Test
    public void testGlobalEnqueueAdditionalThreads() {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.GLOBAL, this::mockFunction);

        assertEquals(1, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(1, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        assertEquals(numData + additionalData, pool.getWorkQueues()[0].size());
    }

    @Test
    public void testRoundRobinEnqueueAdditionalThreads() {
        int numThreads = 3;
        int numData = 3000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.ROUND_ROBIN, this::mockFunction);

        assertEquals(numThreads, pool.getNumWorkQueues());

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(numThreads + additionalThreads, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        int cnt1 = 0;
        int cnt2 = 0;

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            if (queue.size() == numData / numThreads + additionalData / (numThreads + additionalThreads)) {
                cnt1++;
            } else if (queue.size() == additionalData / (numThreads + additionalThreads)) {
                cnt2++;
            }
        }

        assertEquals(cnt1, numThreads);
        assertEquals(cnt2, additionalThreads);
    }

    @Test
    public void testBroadcastDequeueAdditionalThreads() throws InterruptedException {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.BROADCAST, this::mockFunction);

        pool.execute();

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(numThreads + additionalThreads, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        TimeUnit.SECONDS.sleep(3);

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            assertEquals(0, queue.size());
        }
    }

    @Test
    public void testGlobalDequeueAdditionalThreads() throws InterruptedException {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.GLOBAL, this::mockFunction);

        pool.execute();

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(1, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        TimeUnit.SECONDS.sleep(3);

        assertEquals(0, pool.getWorkQueues()[0].size());
    }

    @Test
    public void testRoundRobinDequeueAdditionalThreads() throws InterruptedException {
        int numThreads = 3;
        int numData = 1000;
        pool = new ThreadPool<>(numThreads, ThreadPool.QueuePolicy.ROUND_ROBIN, this::mockFunction);

        pool.execute();

        for (int i = 0; i < numData; i++) {
            pool.enqueue(i);
        }

        int additionalThreads = 2;
        int additionalData = 500;
        pool.createMoreThreads(additionalThreads);

        assertEquals(numThreads + additionalThreads, pool.getNumWorkQueues());

        for (int i = 0; i < additionalData; i++) {
            pool.enqueue(i);
        }

        TimeUnit.SECONDS.sleep(3);

        for (EventfulQueue<Integer> queue : pool.getWorkQueues()) {
            assertEquals(0, queue.size());
        }
    }
}

File: example-crs-webservice/crs-userspace/libs/libMSA/go/thread/threadpool_test.go
package thread

import (
	"testing"
	"time"
)

func mockFunction(data any) {
	time.Sleep(10 * time.Microsecond) // Simulates some processing
}

func TestBroadcastEnqueue(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, BROADCAST, mockFunction)
	
	if len(pool.workQueues) != numThreads {
		t.Errorf("Expected %d work queues, got %d", numThreads, len(pool.workQueues))
	}

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	for _, queue := range pool.workQueues {
		if queue.queue.Len() != 1000 {
			t.Errorf("Expected queue size %d, got %d", numData, queue.queue.Len())
		}
	}
}

func TestGlobalEnqueue(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, GLOBAL, mockFunction)
	
	if len(pool.workQueues) != 1 {
		t.Errorf("Expected 1 work queue, got %d", len(pool.workQueues))
	}

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	if pool.workQueues[0].queue.Len() != numData {
		t.Errorf("Expected queue size %d, got %d", numData, pool.workQueues[0].queue.Len())
	}
}

func TestRoundRobinEnqueue(t *testing.T) {
	numThreads := 3
	numData := 3000
	pool := NewThreadPool(numThreads, ROUND_ROBIN, mockFunction)
	
	if len(pool.workQueues) != numThreads {
		t.Errorf("Expected %d work queues, got %d", numThreads, len(pool.workQueues))
	}

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	for _, queue := range pool.workQueues {
		expectedSize := numData / numThreads
		if queue.queue.Len() != expectedSize {
			t.Errorf("Expected queue size %d, got %d", expectedSize, queue.queue.Len())
		}
	}
}

func TestBroadcastDequeue(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, BROADCAST, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	for _, queue := range pool.workQueues {
		if queue.queue.Len() != 0 {
			t.Errorf("Expected queue size 0 after processing, got %d", queue.queue.Len())
		}
	}
}

func TestGlobalDequeue(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, GLOBAL, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	if pool.workQueues[0].queue.Len() != 0 {
		t.Errorf("Expected queue size 0 after processing, got %d", pool.workQueues[0].queue.Len())
	}
}

func TestRoundRobinDequeue(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, ROUND_ROBIN, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	for _, queue := range pool.workQueues {
		if queue.queue.Len() != 0 {
			t.Errorf("Expected queue size 0 after processing, got %d", queue.queue.Len())
		}
	}
}

func TestBroadcastEnqueueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, BROADCAST, mockFunction)

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	if len(pool.workQueues) != numThreads+additionalThreads {
		t.Errorf("Expected %d work queues, got %d", numThreads+additionalThreads, len(pool.workQueues))
	}

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	cnt1 := 0
	cnt2 := 0
	for i, queue := range pool.workQueues {
		queueSize := queue.queue.Len()
		if i < numThreads && queueSize == numData+additionalData {
			cnt1++
		} else if i >= numThreads && queueSize == additionalData {
			cnt2++
		}
	}

	if cnt1 != numThreads || cnt2 != additionalThreads {
		t.Errorf("Expected %d queues with size %d, got %d; expected %d queues with size %d, got %d",
			numThreads, numData+additionalData, cnt1, additionalThreads, additionalData, cnt2)
	}
}

func TestGlobalEnqueueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, GLOBAL, mockFunction)

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	if len(pool.workQueues) != 1 {
		t.Errorf("Expected 1 work queue, got %d", len(pool.workQueues))
	}

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	expectedSize := numData + additionalData
	if pool.workQueues[0].queue.Len() != expectedSize {
		t.Errorf("Expected queue size %d, got %d", expectedSize, pool.workQueues[0].queue.Len())
	}
}

func TestRoundRobinEnqueueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 3000
	pool := NewThreadPool(numThreads, ROUND_ROBIN, mockFunction)

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	if len(pool.workQueues) != numThreads+additionalThreads {
		t.Errorf("Expected %d work queues, got %d", numThreads+additionalThreads, len(pool.workQueues))
	}

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	cnt1 := 0
	cnt2 := 0
	for i, queue := range pool.workQueues {
		expectedSize := numData/numThreads + additionalData/(numThreads+additionalThreads)
		if queue.queue.Len() == expectedSize {
			cnt1++
		} else if i >= numThreads && queue.queue.Len() == additionalData/(numThreads+additionalThreads) {
			cnt2++
		}
	}

	if cnt1 != numThreads || cnt2 != additionalThreads {
		t.Errorf("Expected %d queues with size %d, got %d; expected %d queues with size %d, got %d",
			numThreads, numData/numThreads+additionalData/(numThreads+additionalThreads), cnt1,
			additionalThreads, additionalData/(numThreads+additionalThreads), cnt2)
	}
}

func TestBroadcastDequeueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, BROADCAST, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	for _, queue := range pool.workQueues {
		if queue.queue.Len() != 0 {
			t.Errorf("Expected queue size 0 after processing, got %d", queue.queue.Len())
		}
	}
}

func TestGlobalDequeueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, GLOBAL, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	if pool.workQueues[0].queue.Len() != 0 {
		t.Errorf("Expected queue size 0 after processing, got %d", pool.workQueues[0].queue.Len())
	}
}

func TestRoundRobinDequeueAdditionalThreads(t *testing.T) {
	numThreads := 3
	numData := 1000
	pool := NewThreadPool(numThreads, ROUND_ROBIN, mockFunction)
	pool.Start()

	for i := 0; i < numData; i++ {
		pool.Enqueue(i)
	}

	additionalThreads := 2
	additionalData := 500
	pool.AddThreads(additionalThreads)

	for i := 0; i < additionalData; i++ {
		pool.Enqueue(i)
	}

	time.Sleep(3 * time.Second) // Wait for processing to finish

	for _, queue := range pool.workQueues {
		if queue.queue.Len() != 0 {
			t.Errorf("Expected queue size 0 after processing, got %d", queue.queue.Len())
		}
	}
}